core/src/main/java/io/questdb/cutlass/line/tcp/TableUpdateDetails.java

```diff
@@ -318,16 +320,84 @@ int getColumnCount() {
             return columnCount;
         }
 
+        StringSink getTempSink() {
+            return tempSink;
+        }
+
         int getColumnIndex(DirectByteCharSequence colName) {
-            final int colIndex = columnIndexByNameUtf8.get(colName);
-            if (colIndex != CharSequenceIntHashMap.NO_ENTRY_VALUE) {
+            int colIndex = columnIndexByNameUtf8.get(colName);
+            if (colIndex == COLUMN_INDEX_NOT_CACHED) {
+                colIndex = getColumnIndexFromReader(colName);
+                if (colIndex < 0) {
+                    unresolvedColumnFlag = true;
+                }
+            }
+            return colIndex;
+        }
+
+        private int getColumnIndexFromReader(DirectByteCharSequence colName) {
+            try (TableReader reader = engine.getReader(AllowAllCairoSecurityContext.INSTANCE, tableNameUtf16)) {
+                TableReaderMetadata metadata = reader.getMetadata();
+                tempSink.clear();
+                if (!Chars.utf8Decode(colName.getLo(), colName.getHi(), tempSink)) {
+                    throw CairoException.instance(0).put("invalid UTF8 in value for ").put(colName);
+                }
+                int colIndex = metadata.getColumnIndexQuiet(tempSink);
+                columnIndexByNameUtf8.put(colName.toString(), colIndex);
+                updateColumTypeCache(metadata, colIndex);
                 return colIndex;
             }
-            return populateCacheAndGetColumnIndex(colName);
+        }
+
+        void updateColumnIndexCache() {
+            if (!unresolvedColumnFlag) {
+                return;
+            }
+
+            try (TableReader reader = engine.getReader(AllowAllCairoSecurityContext.INSTANCE, tableNameUtf16)) {
+                TableReaderMetadata metadata = reader.getMetadata();
+                unresolvedColumnFlag = false;
+                boolean resolvedNewColumn = false;
```

Comment:
there is peculiar state management of `unresolvedColumnFlag` here. It would be good if it was simpler. What could happen here is that if `TableReader` constructor were to fail, the `true` value in `unresolvedColumnFlag` would get stuck.

By simpler state management I meant logically decoupling `unresolvedColumnFlag` from reader availability and perhaps tying it up with starting of new event processing.

Reply:
changed the logic to stop looking up new column names after we have found the first unresolved column name.
we will start looking new column names up only on the next event.
this change improves state handling.